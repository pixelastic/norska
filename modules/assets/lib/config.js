module.exports = {
  files: [
    // Static files
    '**/*.{html,txt}',
    // Images
    '**/*.{ico,jpg,jpeg,gif,png,svg}',
    // Fonts
    '**/*.{eot,otf,ttf,woff,woff2}',
    // Documents
    '**/*.pdf',
    // Netlify
    '_redirects',
    '_headers',
    'netlify.toml',
  ],
};
