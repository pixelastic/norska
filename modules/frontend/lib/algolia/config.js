module.exports = {
  options: {
    routerIgnore: [], // List of parameters to ignore in the router (will not be added to the URL)
  },
};
