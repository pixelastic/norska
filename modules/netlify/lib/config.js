module.exports = {
  deploy: {
    // Deploy again if any of those files were changed since the last deploy:
    // - All path are relative to the repo root
    // - <projectRoot> is where the package.json holding norska is
    // - <from> is the source folder, relative to the <projectRoot>
    files: [
      'lambda/**/*',
      'netlify.toml',
      '.nvmrc',
      '<projectRoot>/norska.config.js',
      '<projectRoot>/tailwind.config.js',
      '<from>/**/*',
    ],
    // If any of those keys in package.json  has been modified since the last
    // deploy, we deploy again
    keys: ['dependencies', 'scripts.build:prod'],
  },
};
