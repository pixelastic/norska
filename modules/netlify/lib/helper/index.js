const NetlifyAPI = require('netlify');
const config = require('norska-config');
const exists = require('firost/exists');
const readJson = require('firost/readJson');
const got = require('golgoth/got');
const _ = require('golgoth/lodash');
module.exports = {
  /**
   * Returns the Netlify token saved in ENV
   * @returns {string} The Netlify token
   **/
  token() {
    return this.getEnvVar('NETLIFY_AUTH_TOKEN');
  },
  /**
   * Check if a Netlify token is available
   * @returns {boolean} True if a token is defined
   **/
  hasToken() {
    return !!this.token();
  },
  /**
   * Returns a Netlify API Client
   * Always returns the same instance
   * @returns {object} NetlifyAPI instance
   **/
  apiClient() {
    if (!this.__apiClient) {
      const token = this.token();
      this.__apiClient = new NetlifyAPI(token);
    }
    return this.__apiClient;
  },
  /**
   * Check if the current code is running on Netlify servers
   * @returns {boolean} True if running remotely, false otherwise
   **/
  isRunningRemotely() {
    return !!this.getEnvVar('NETLIFY');
  },
  /**
   * Returns the current siteId
   * @returns {string} Netlify Site ID
   **/
  async siteId() {
    if (this.__siteId) {
      return this.__siteId;
    }

    // Remotely: saved as SITE_ID
    if (this.isRunningRemotely()) {
      return this.getEnvVar('SITE_ID');
    }

    // Locally: Saved in .netlify/state.json
    return await this.localSiteId();
  },
  /**
   * Check if the current repo is linked to a remote site
   * @returns {boolean} True if linked
   **/
  async isLinkedLocally() {
    return !!(await this.localSiteId());
  },
  /**
   * Returns the siteId from the local filesystem
   * @returns {string} Netlify siteId
   **/
  async localSiteId() {
    const netlifyStateFile = config.rootPath('.netlify/state.json');
    if (!(await exists(netlifyStateFile))) {
      return false;
    }
    const netlifyState = await readJson(netlifyStateFile);
    return netlifyState.siteId;
  },
  /**
   * Returns the current Netlify site name
   * @returns {string} Site name
   **/
  async siteName() {
    const siteId = await this.siteId();
    const client = this.apiClient();
    const response = await client.getSite({ site_id: siteId });
    return response.name;
  },
  /**
   * Returns an ENV var by name.
   * @param {string} key Name of the var
   * @returns {string} Value of the var
   **/
  getEnvVar(key) {
    return process.env[key];
  },
  /**
   * Make a raw API call to the Netlify API without using their API client
   * because some endpoints are not implemented
   * @param {string} apiPath Path of the api, after the api/v1 part
   * @param {object} userOptions Options to pass to the internal got call
   * @returns {object} Parsed JSON response or raw response
   **/
  async rawApiCall(apiPath, userOptions = {}) {
    const token = this.token();
    const apiUrl = `https://api.netlify.com/api/v1/${apiPath}`;
    const defaultOptions = {
      method: 'get',
      headers: {
        authorization: `Bearer ${token}`,
      },
    };
    const options = _.merge({}, defaultOptions, userOptions);
    const response = await got(apiUrl, options);
    try {
      return JSON.parse(response.body);
    } catch (err) {
      return response.body;
    }
  },
};
