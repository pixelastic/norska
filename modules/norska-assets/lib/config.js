module.exports = {
  files: [
    // Static files
    '**/*.{html,txt}',
    // Images
    '**/*.{ico,jpg,gif,png,svg}',
    // Fonts
    '**/*.{eot,otf,ttf,woff,woff2}',
    // Documents
    '**/*.pdf',
    // Netlify
    '_redirects',
    '_headers',
    'netlify.toml',
  ],
};
