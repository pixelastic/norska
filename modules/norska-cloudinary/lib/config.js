module.exports = {
  enable: true, // Set to false to disable the module
  bucketName: null, // bucketName used in urls
};
