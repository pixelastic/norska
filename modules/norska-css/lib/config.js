module.exports = {
  input: 'style.css',
};
