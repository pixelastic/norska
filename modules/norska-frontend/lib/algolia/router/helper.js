const credentials = require('../credentials.js');
module.exports = {
  metaKeys: ['page', 'query'],
  indexKeys: ['sortBy', 'index'],
  filterKeys: ['refinementList', 'range'],
  rangePattern: /^{(.*)}/,
  refinementListPattern: /^\[(.*)\]/,
  /**
   * Converts a key and value pair to a string suitable to be put in the URL
   * @param {*} value Value of the filters
   * @param {string} key Key name
   * @returns {string} URL part
   **/
  keyValueToString(value, key) {
    const isMeta = this.metaKeys.includes(key);
    const isIndex = this.indexKeys.includes(key);

    if (isMeta || isIndex) {
      return `${key}:${encodeURIComponent(value)}`;
    }

    const isFilter = Array.isArray(value);

    // Filters
    if (isFilter) {
      const encodedValue = value.map(encodeURIComponent);
      const joinedValue = encodedValue.sort().join(',');
      return `${key}:[${joinedValue}]`;
    }

    // Range
    const isRange = value.includes(':');
    if (isRange) {
      const urlValue = value.replace(':', ',');
      return `${key}:{${urlValue}}`;
    }

    return 'TODO';
  },
  /**
   * Converts a key:value hash string into a filter object
   * @param {string} input key:value hash string
   * @returns {object} Filter object
   **/
  stringToFilterObject(input) {
    const [rawKey, rawValue] = input.split(':');
    let key = rawKey;
    let value = rawValue;

    // Stop if a filter than changes the index
    const isIndex = this.indexKeys.includes(key);
    if (isIndex) {
      return {};
    }

    // Pages must be converted to number
    const isPage = key === 'page';
    if (isPage) {
      value = parseInt(value);
    }

    // Meta keys are at the root
    const isMeta = this.metaKeys.includes(key);
    if (isMeta) {
      return {
        [key]: decodeURIComponent(value),
      };
    }

    // Refinement list
    const isFilter = this.refinementListPattern.test(value);
    if (isFilter) {
      const capture = this.refinementListPattern.exec(value);
      value = capture[1]
        .split(',')
        .map(decodeURIComponent)
        .sort();
      return {
        refinementList: {
          [key]: value,
        },
      };
    }

    // Range
    const isRange = this.rangePattern.test(value);
    if (isRange) {
      const capture = this.rangePattern.exec(value);
      value = capture[1].replace(',', ':');
      return {
        range: {
          [key]: value,
        },
      };
    }
  },
  /**
   * Return the name of the index base on the location hash
   * Default to base index, unless index: is set, or sortBy: is set
   * @param {string} locationHash location hash to decode
   * @returns {string} name of the index
   */
  indexNameFromLocationHash(locationHash) {
    const stringFilters = locationHash.replace(/^#/, '').split('/');
    const filters = {};
    stringFilters.forEach(filter => {
      const [key, value] = filter.split(':');
      filters[key] = value;
    });

    const baseIndex = this.indexName();
    const sortByKey = filters.sortBy;
    const indexKey = filters.index;
    if (sortByKey) {
      return `${baseIndex}_${sortByKey}`;
    }
    if (indexKey) {
      return indexKey;
    }
    return baseIndex;
  },
  /**
   * Return the current url
   * @returns {string} Current url, without query string nor hash
   **/
  currentUrl() {
    const fullUrl = `${location.origin}${location.pathname}`;
    return fullUrl.replace(/index\.html$/, '');
  },
  /**
   * Returns the index name
   * @returns {string} Name of the main index
   **/
  indexName() {
    return credentials.indexName();
  },
  /**
   * Checks if a given index name seems to be a replica
   * Replicas start with the base index name as prefix
   * @param {string} indexName name of the index to test
   * @returns {boolean} true if seems like a replica
   **/
  isReplica(indexName) {
    const baseIndexName = this.indexName();
    return indexName.startsWith(baseIndexName);
  },
  /**
   * Return the short replica name (minus the base index prefix)
   * @param {string} indexName name of the replica index
   * @returns {string} short name of the replica
   **/
  replicaShortname(indexName) {
    const baseIndexName = this.indexName();
    const regexp = new RegExp(`^${baseIndexName}_`);
    return indexName.replace(regexp, '');
  },
};
