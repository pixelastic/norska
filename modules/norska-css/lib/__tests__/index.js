import module from '../index';
import config from 'norska-config';
import helper from 'norska-helper';
import firost from 'firost';

describe('norska-css', () => {
  beforeEach(async () => {
    await config.init({
      from: './fixtures/src',
      to: './tmp/norska-css',
    });
  });
  describe('getPlugins', () => {
    beforeEach(() => {
      jest.spyOn(module, '__pluginImport').mockReturnValue('pluginImport');
      jest.spyOn(module, '__pluginNested').mockReturnValue('pluginNested');
      jest.spyOn(module, '__pluginPurge').mockReturnValue('pluginPurge');
      jest.spyOn(module, '__pluginClean').mockReturnValue('pluginClean');
      jest
        .spyOn(module, '__pluginAutoprefixer')
        .mockReturnValue('pluginAutoprefixer');
    });
    it('should contain 2 plugins', () => {
      const actual = module.getPlugins();

      expect(actual).toEqual(['pluginImport', 'pluginNested']);
    });
    describe('in production', () => {
      beforeEach(() => {
        jest.spyOn(helper, 'isProduction').mockReturnValue(true);
      });
      it('should contain 5 plugins', () => {
        const actual = module.getPlugins();

        expect(actual).toEqual([
          'pluginImport',
          'pluginNested',
          'pluginPurge',
          'pluginAutoprefixer',
          'pluginClean',
        ]);
      });
    });
  });
  describe('getCompiler', () => {
    it('should return the postcss().process method, correctly bound', () => {
      jest.spyOn(module, 'getPlugins').mockReturnValue('my plugins');
      jest.spyOn(module, '__postcss').mockImplementation(function(plugins) {
        return {
          plugins,
          process: jest.fn().mockImplementation(function() {
            return this.plugins;
          }),
        };
      });

      const actual = module.getCompiler();

      expect(actual()).toEqual('my plugins');
    });

    // const plugins = this.getPlugins();

    // const postcssInstance = this.__postcss(plugins);
    // return _.bind(postcssInstance.process, postcssInstance);
  });
  describe('compile', () => {
    beforeEach(() => {
      jest.spyOn(firost, 'read').mockReturnValue();
      jest.spyOn(firost, 'write').mockReturnValue();
    });
    it('should fail if file is not in the source folder', async () => {
      jest.spyOn(helper, 'consoleWarn').mockReturnValue();
      const input = '/nope/foo.css';

      const actual = await module.compile(input);

      expect(actual).toEqual(false);
      expect(helper.consoleWarn).toHaveBeenCalled();
    });
    it('should call the compiler with the raw content', async () => {
      jest.spyOn(firost, 'read').mockReturnValue('css content');
      const mockCompiler = jest.fn();
      jest.spyOn(module, 'getCompiler').mockReturnValue(mockCompiler);

      await module.compile('style.css');

      expect(mockCompiler).toHaveBeenCalledWith(
        'css content',
        expect.anything()
      );
    });
    it('should call the compiler with from as the source', async () => {
      jest.spyOn(firost, 'read').mockReturnValue('css content');
      const mockCompiler = jest.fn();
      jest.spyOn(module, 'getCompiler').mockReturnValue(mockCompiler);

      await module.compile('style.css');

      expect(mockCompiler).toHaveBeenCalledWith(
        expect.anything(),
        expect.objectContaining({ from: config.fromPath('style.css') })
      );
    });
    it('should write the .css key result to file', async () => {
      const mockCompiler = jest.fn().mockImplementation(async () => {
        return { css: 'compiled content' };
      });
      jest.spyOn(module, 'getCompiler').mockReturnValue(mockCompiler);

      await module.compile('style.css');

      expect(firost.write).toHaveBeenCalledWith(
        'compiled content',
        expect.anything()
      );
    });
    it('should write the compiled css to destination file', async () => {
      const mockCompiler = jest.fn();
      jest.spyOn(module, 'getCompiler').mockReturnValue(mockCompiler);

      await module.compile('style.css');

      expect(firost.write).toHaveBeenCalledWith(
        undefined,
        config.toPath('style.css')
      );
    });
  });
  describe('run', () => {
    describe('in development', () => {
      it('should compile the input file', async () => {
        jest.spyOn(helper, 'isProduction').mockReturnValue(false);
        config.init({
          from: './fixtures/src',
          to: './tmp/norska-css',
          css: module.defaultConfig(),
        });
        await firost.emptyDir('./tmp/norska-css');
        await module.run();

        const actual = await firost.read(config.toPath('style.css'));

        expect(actual).toMatchSnapshot();
      });
    });
    describe('in production', () => {
      it('should compile the input file', async () => {
        jest.spyOn(helper, 'isProduction').mockReturnValue(true);
        config.init({
          from: './fixtures/src',
          to: './tmp/norska-css',
          css: module.defaultConfig(),
        });
        await firost.emptyDir('./tmp/norska-css');
        await firost.write(
          '<p class="context"><b>foo</b></p>',
          config.toPath('index.html')
        );
        await module.run();

        const actual = await firost.read(config.toPath('style.css'));

        expect(actual).toMatchSnapshot();
      });
    });
  });
});
