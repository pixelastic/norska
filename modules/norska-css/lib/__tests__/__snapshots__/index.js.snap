// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`norska-css run in development should compile the input file 1`] = `
"/*! This is a special comment that should be removed in production */
/* @import statements should be included inline */
.context { background: blue; }
/* nested syntax should be flatten inline */
.context b {
    /* Autoprefixer should add vendor-prefixes in production */
    user-select: none;
  }
/* This class is not available in the html, so should be removed in the
 * production build */
.not-present { color: red; }"
`;

exports[`norska-css run in production should compile the input file 1`] = `".context{background:#00f}.context b{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}"`;
